contracts/protocol/tokenization/PYieldToken.sol

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.10;
+
+import {IPool} from "../../interfaces/IPool.sol";
+import {PToken} from "./PToken.sol";
+import {WadRayMath} from "../libraries/math/WadRayMath.sol";
+import {XTokenType} from "../../interfaces/IXTokenType.sol";
+import {SafeCast} from "../../dependencies/openzeppelin/contracts/SafeCast.sol";
+import {Errors} from "../libraries/helpers/Errors.sol";
+import {IERC20} from "../../dependencies/openzeppelin/contracts/IERC20.sol";
+import {GPv2SafeERC20} from "../../dependencies/gnosis/contracts/GPv2SafeERC20.sol";
+import {IYieldInfo} from "../../interfaces/IYieldInfo.sol";
+import {IAutoYieldApe} from "../../interfaces/IAutoYieldApe.sol";
+
+/**
+ * @title Rebasing PToken
+ *
+ * @notice Implementation of the interest bearing token for the ParaSpace protocol
+ */
+contract PYieldToken is PToken {
+    using WadRayMath for uint256;
+    using SafeCast for uint256;
+    using GPv2SafeERC20 for IERC20;
+
+    uint256 internal constant RAY = 1e27;
+    mapping(address => uint256) private _userYieldIndex;
+    mapping(address => uint256) private _userPendingYield;
+    mapping(address => uint256) private _userLockFeeAmount;
+
+    constructor(IPool pool) PToken(pool) {
+        //intentionally empty
+    }
+
+    function mint(
+        address caller,
+        address onBehalfOf,
+        uint256 amount,
+        uint256 index
+    ) external override onlyPool returns (bool) {
+        _updateUserIndex(onBehalfOf, int256(amount));
+
+        return _mintScaled(caller, onBehalfOf, amount, index);
+    }
+
+    function burn(
+        address from,
+        address receiverOfUnderlying,
+        uint256 amount,
+        uint256 index
+    ) external override onlyPool {
+        _updateUserIndex(from, -int256(amount));
+
+        _burnScaled(from, receiverOfUnderlying, amount, index);
+        if (receiverOfUnderlying != address(this)) {
+            IERC20(_underlyingAsset).safeTransfer(receiverOfUnderlying, amount);
+        }
+    }
+
+    function _transfer(
+        address from,
+        address to,
+        uint256 amount,
+        bool validate
+    ) internal override {
+        _updateUserIndex(from, -int256(amount));
+        _updateUserIndex(to, int256(amount));
+
+        super._transfer(from, to, amount, validate);
+    }
+
+    function _yieldAmount(address account)
+        internal
+        view
+        returns (uint256, uint256)
+    {
+        uint256 userBalance = balanceOf(account);
+        uint256 pendingYield = _userPendingYield[account];
+        uint256 lockedYield = 0;
+        if (userBalance > 0) {
+            uint256 userIndex = _userYieldIndex[account];
+            (uint256 lastAccruedIndex, uint256 latestYieldIndex) = IYieldInfo(
+                _underlyingAsset
+            ).yieldIndex();
+            uint256 indexDiff = latestYieldIndex - userIndex;
```

Comment:
I think we need some comments to expain the following calcualtion.

Reply:
Consider give it a meaningful name. It would be better to represent the business meaning rather than the calculation meaning.